### PR TITLE
Add initial tests for the "iterator-sequencing" proposal

### PR DESCRIPTION
**Pull request by anba**
Tests are based on existing tests in "test/built-ins/Iterator".

### CODE REVIEW

File: test/built-ins/Iterator/concat/next-method-returns-non-object.js
Comment (ljharb) [inline test/built-ins/Iterator/concat/next-method-returns-non-object.js:36]: in my polyfill, `.next().next()` throws instead. this is a reminder for myself

File: test/built-ins/Iterator/concat/next-method-returns-throwing-done.js
Comment (ljharb) [inline test/built-ins/Iterator/concat/next-method-returns-throwing-done.js:44]: here too

File: test/built-ins/Iterator/concat/next-method-returns-throwing-value.js
Comment (ljharb) [inline test/built-ins/Iterator/concat/next-method-returns-throwing-value.js:44]: and here

File: test/built-ins/Iterator/concat/next-method-throws.js
Comment (ljharb) [inline test/built-ins/Iterator/concat/next-method-throws.js:36]: and here

File: test/built-ins/Iterator/concat/return-is-forwarded.js
Comment (ljharb) [inline test/built-ins/Iterator/concat/return-is-forwarded.js:58]: my polyfill reports 2 here

File: test/built-ins/Iterator/concat/throws-typeerror-when-generator-is-running-next.js
Comment (ljharb) [inline test/built-ins/Iterator/concat/throws-typeerror-when-generator-is-running-next.js:44]: also here

File: test/built-ins/Iterator/concat/throws-typeerror-when-generator-is-running-return.js
Comment (ljharb) [inline test/built-ins/Iterator/concat/throws-typeerror-when-generator-is-running-return.js:53]: lol this one blew the stack of my polyfill at 864, clearly i have a bug in this one

**Review by ljharb — COMMENTED** [review]
Almost all of these tests pass verbatim on my polyfill, which is great!

I've commented the ones that don't - I assume they represent bugs in my polyfill, but I'll do a deep dive on the spec text to make sure.

File: test/built-ins/Iterator/concat/throws-typeerror-when-generator-is-running-return.js
Comment (ljharb) [inline test/built-ins/Iterator/concat/throws-typeerror-when-generator-is-running-return.js:50]: fixed the other two issues, but now mine doesn't throw here.

Comment (michaelficarra) [reply test/built-ins/Iterator/concat/throws-typeerror-when-generator-is-running-return.js:50]: 1. this line calls `%IteratorHelperPrototype%.return`
1. `%IteratorHelperPrototype%.return` step 6 calls `GeneratorResumeAbrupt`
1. `GeneratorResumeAbrupt` step 8 sets `[[GeneratorState]]` to executing
1. `GeneratorResumeAbrupt` step 10 resumes the underlying generator
1. `Iterator.concat` step 3.a.v.3.a, Yield produces the return completion from `%IteratorHelperPrototype%.return` step 5
1. `Iterator.concat` step 3.a.v.3.b.i calls `IteratorClose`
1. `IteratorClose` step 4.c calls `testIterator.return`
1. line 32 calls `%IteratorHelperPrototype%.return`
1. `%IteratorHelperPrototype%.return` step 6 calls `GeneratorResumeAbrupt`
1. `GeneratorResumeAbrupt` step 1 calls `GeneratorValidate`
1. `GeneratorValidate` step 6 throws

Comment (anba) [reply test/built-ins/Iterator/concat/throws-typeerror-when-generator-is-running-return.js:50]: https://github.com/es-shims/iterator-helpers/blob/dd80b2f1718092219e94e24afdc1737f062e3df4/aos/GeneratorResumeAbrupt.js#L34-L37 seems to be missing to implement GeneratorResumeAbrupt, step 8 where `generator.[[GeneratorState]]` is set to `executing`.

Comment (ljharb) [reply test/built-ins/Iterator/concat/throws-typeerror-when-generator-is-running-return.js:50]: I've pushed up some local changes I had, so now: if i comment out/remove this (incorrect) block https://github.com/es-shims/iterator-helpers/blob/temp/aos/GeneratorResumeAbrupt.js#L39-L42

then i go from 1 failing test to 9 (from `node test`). I'm pretty convinced this is a polyfill problem, though, because I'm not using generators and I've made my own quasi-state machine that surely hits some edge cases.

File: test/built-ins/Iterator/concat/throws-typeerror-when-iterable-not-an-object.js
Comment (michaelficarra) [inline test/built-ins/Iterator/concat/throws-typeerror-when-iterable-not-an-object.js:13]: Relatedly, the primitive object wrappers (`new Boolean(true)`, etc) should pass this step. We should have a test for that.

File: test/built-ins/Iterator/concat/result-is-iterator.js
Comment (michaelficarra) [inline test/built-ins/Iterator/concat/result-is-iterator.js:13]: We should make this same assertion when passed 1 iterator that doesn't inherit from Iterator. I could imagine an implementation trying to make an optimisation for the 1-iterator case where they pass the argument right through.

File: test/built-ins/Iterator/concat/many-arguments.js
Comment (michaelficarra) [inline test/built-ins/Iterator/concat/many-arguments.js:33]: As the champion of `Array.prototype.flat`, it breaks my heart to see this.

```suggestion
let array = iterables.flat();
```

File: test/built-ins/Iterator/concat/fresh-iterator-result.js
Comment (michaelficarra) [inline test/built-ins/Iterator/concat/fresh-iterator-result.js:49]: This seems unnecessarily wasteful. I've opened an issue to reconsider this behaviour: https://github.com/tc39/proposal-iterator-sequencing/issues/17

**Review by michaelficarra — APPROVED** [review]
LGTM otherwise! Thanks for writing these, @anba.

File: test/built-ins/Iterator/concat/many-arguments.js
Comment (ptomato) [reply test/built-ins/Iterator/concat/many-arguments.js:33]: We try to avoid newer ES language features in tests if they're not the feature being tested - it's not a prohibition and many tests do it anyway, but we generally don't go out of our way to add them.

**Comment by michaelficarra** [comment]
I pushed up a couple of commits. This LGTM now, modulo the normative question which we can resolve at the upcoming plenary.

File: test/built-ins/Iterator/concat/fresh-iterator-result.js
Comment (michaelficarra) [reply test/built-ins/Iterator/concat/fresh-iterator-result.js]: See https://github.com/tc39/proposal-iterator-sequencing/pull/18#issuecomment-2512825730. I've updated the test.

**Review by ptomato — APPROVED** [review]
As discussed in the maintainers chat, merging this on the strength of Jordan's and Michael's reviews.

File: test/built-ins/Iterator/concat/return-is-not-forwarded-before-initial-start.js
Comment (bnoordhuis) [inline test/built-ins/Iterator/concat/return-is-not-forwarded-before-initial-start.js:32]: Implementer here. I'm confused by this test, can someone explain?

I understand why the first return() should not forward, but the second one should, right? And that makes the test fail because the return method throws.

Is return-before-next a terminal state? But then next() should not be invoked either, it seems like? But if that's the case, I would expect its implementation in this file to throw.

Comment (anba) [reply test/built-ins/Iterator/concat/return-is-not-forwarded-before-initial-start.js:32]: 1. `Iterator.concat` calls `CreateIteratorFromClosure` which in turn calls `GeneratorStart`. `[[GeneratorState]]` is still "suspended-start" at this point. This finishes initialising the iterator and control returns to the caller.
2. When now calling `%IteratorHelperPrototype%.return()`, `[[GeneratorState]]` is still "suspended-start" and we transition `[[GeneratorState]]` to "completed". `IteratorClose` is called for all opened underlying iterators, but at this point no iterator has yet been opened.
3. When then calling `%IteratorHelperPrototype%.next()`, nothing should happen, because `GeneratorResume` directly returns when `[[GeneratorState]]` is "completed".
4. And when calling `%IteratorHelperPrototype%.return()` after `next()`, still no user-code will be called, because `GeneratorResumeAbrupt` also directly returns when `[[GeneratorState]]` is "completed".

Comment (bnoordhuis) [reply test/built-ins/Iterator/concat/return-is-not-forwarded-before-initial-start.js:32]: Thanks! Does that mean the next() method here could/should throw Test262Error, like the return() method does?

Comment (anba) [reply test/built-ins/Iterator/concat/return-is-not-forwarded-before-initial-start.js:32]: Yes, `testIterator.next` can be changed to throw a `Test262Error` without changing the test semantics.

Comment (bnoordhuis) [reply test/built-ins/Iterator/concat/return-is-not-forwarded-before-initial-start.js:32]: Thanks, I've opened https://github.com/tc39/test262/pull/4585 to make that more explicit.

**Pull request merged**
